src/lib/CofThy.ml

```diff
@@ -1,209 +1,278 @@
-    | `Inconsistent -> `Inconsistent, []
-    | `Consistent {classes; true_vars; irreducible_joins} ->
-      `Consistent {classes; true_vars}, List.map Cof.join irreducible_joins
+    | Cof.Cof phi ->
+      begin
+        match phi with
+        | Cof.Eq (r, s) ->
+          test_eq thy' (r, s)
+        | Cof.Join phis ->
+          List.exists (test thy') phis
+        | Cof.Meet phis ->
+          List.for_all (test thy') phis
+      end
+    | Cof.Var v ->
+      test_var thy' v
+
+  let left_invert_under_cofs ~zero ~seq (thy : t) cofs cont =
+    match split thy cofs with
+    | [] -> zero
+    | [thy] -> cont thy
+    | thys -> seq cont thys
+end
 
-  let assemble_thy alg_thy =
-    assume @@ disj_envelope alg_thy
+module Disj =
+struct
+  type t = disj_thy
 
-  let left_invert_under_cofs ~zero ~seq alg_thy phis cont =
-    Search.left_invert ~zero ~seq ~fast_track:(fun _ x -> x ()) (disj_envelope alg_thy) phis @@ fun alg_thy' ->
-    cont @@ `Consistent alg_thy'
+  let init : t = [Alg.init', (VarSet.empty, [])]
+
+  let consistency =
+    function
+    | [] -> `Inconsistent
+    | _ -> `Consistent
+
+  (** [refactor_branches] attempts to identify common parts of the branches
+    * and shrink the labels. Recall that we do not keep the common ancestor
+    * but the paths (as a collection of atomic cofibrations) from the common
+    * ancestor, and thus what are changed here are the paths.
+    *
+    * This optimization seems to be expensive, but it seems to help after
+    * we switched from the persistant tables (using [Hashtbl]) to [Map].
+  *)
+  let refactor_branches cached_branches : t =
+    let common_vars =
+      let go vars0 (_, (vars1, _)) = VarSet.inter vars0 vars1 in
+      match cached_branches with
+      | [] -> VarSet.empty
+      | (_, (vars, _)) :: branches -> List.fold_left go vars branches
+    in
+    (* The following is an approximation of checking whether some equation is useful.
+     * It does not kill every "useless" cofibration. Here is one example:
+     *
+     * branch 1: r=0
+     * branch 2: r=i, i=0
+     *
+     * r=0 will be factored out, but then a=0 should also be removed. Here is a more
+     * complicated example:
+     *
+     * branch 1: r=i, i=0
+     * branch 2: r=j, j=0
+     *
+     * Both i=0 and j=0 should be factored out, but the following code is not smart
+     * enough to detect them. One could consider more aggressive approaches if [CofThy]
+     * becomes the bottleneck again.
+    *)
+    let useful eq = cached_branches |> List.exists @@ fun (thy', _) -> not @@ Alg.test_eq thy' eq in
+    (* revisit all branches and remove all useless ones identified by the simple criterion above. *)
+    cached_branches |> List.map @@ fun (thy', (vars, eqs)) ->
+    thy', (VarSet.diff vars common_vars, List.filter useful eqs)
+
+  (** [split] incorporate a cofibration context and split an already split theory
+    * into more branches. This is similar to [Alg.split] in the spirit, but differs
```

Comment:
I am having trouble following this explanation; could you rephrase or clarify? Is it saying something like:

> `split` adjoins to a disjunctive theory the disjunction of a list of cofibrations, applying certain optimizations beforehand including [XX,YY].

Reply:
No, but I updated the comments.